Fix: function-redefined not triggered for functions with leading underscores (close #9894)

Akshay9715 opened the pull request:
<!--
Thank you for submitting a PR to pylint!

To ease the process of reviewing your PR, do make sure to complete the following boxes.

- [ ] Document your change, if it is a non-trivial one.
  - A maintainer might label the issue ``skip-news`` if the change does not need to be in the changelog.
  - Otherwise, create a news fragment with ``towncrier create <IssueNumber>.<type>`` which will be
    included in the changelog. ``<type>`` can be one of the types defined in `./towncrier.toml`.
    If necessary you can write details or offer examples on how the new change is supposed to work.
  - Generating the doc is done with ``tox -e docs``
- [ ] Relate your change to an issue in the tracker if such an issue exists (Refs #1234, Closes #1234)
- [ ] Write comprehensive commit messages and/or a good description of what the PR does.
- [ ] Keep the change small, separate the consensual changes from the opinionated one.
  Don't hesitate to open multiple PRs if the change requires it. If your review is so
  big it requires to actually plan and allocate time to review, it's more likely
  that it's going to go stale.
- [ ] If you used multiple emails or multiple names when contributing, add your mails
      and preferred name in ``script/.contributors_aliases.json``
-->

## Type of Changes

<!-- Leave the corresponding lines for the applicable type of change: -->

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
| ✓   | :scroll: Docs          |

## Description

<!-- If this PR references an issue without fixing it: -->

Refs #9894

<!-- If this PR fixes an issue, use the following to automatically close when we merge: -->

Closes #9894


#### Fixes a false negative in the `function-redefined` checker.  
* Previously, functions with leading underscores (e.g., `_my_func`) were ignored by the `dummy_variables_rgx` check, so redefining them did not trigger a Pylint warning.  

* This PR updates `basic_error_checker.py` so that `_my_func` and other functions are **no longer skipped** by the dummy variable regex for function definitions.  

* Also includes a unit test in `tests/checkers/base/unittest_basic_error_checker.py` to confirm that `function-redefined` is raised correctly for functions with leading underscores.

Akshay9715 commented on pylint/checkers/base/basic_error_checker.py:
```diff
@@ -573,10 +576,22 @@ def _check_redefinition(
+
+            # Only skip dummy names for non-function nodes (variables, etc.)
             if dummy_variables_rgx and dummy_variables_rgx.match(node.name):
-                return
+                if node.name == "_":
```
The goal of this change is to ensure that `dummy_variables_rgx` only excludes truly “**dummy**” functions (like `_` used intentionally as a placeholder) and not valid user-defined functions that happen to start with an underscore (e.g. `_my_func`).

Previously, any function whose name matched `dummy_variables_rgx` was skipped entirely, which caused false negatives for `_my_func` or `_handle_data`.

So, the fix ensures that only the literal name `_` (a true dummy function) is skipped, while `_my_func` and similar private function names are properly checked for redefinition.

DanielNoord replied:
Yes, but in the linked issue @Pierre-Sassoulas already refers to the fact that potentially we shouldn't be checking the dummy regex to begin with for this check. Honestly, I also don't see a good reason for redefinition although the `singledispatch` one was one I hadn't thought of. It feels kind of weird to me to only partially check it: either the regex should "work" or it should be completely disregarded.

Akshay9715 replied:
That’s a fair point — I agree it would make sense to consistently handle dummy_variables_rgx either entirely or not at all for this checker.

My reasoning for the current approach was to make a minimal, behavior-preserving change that only fixed the immediate false negatives without altering other checkers that might rely on the same regex logic.

But if the consensus is that `function-redefined` should simply ignore dummy_variables_rgx altogether, I can adjust the patch accordingly (and simplify this part).

Should I go ahead and remove the dummy_variables_rgx check completely for this message?

Akshay9715 commented on tests/checkers/base/unittest_basic_error_checker.py:
Got it — I’ll move this to the functional test framework and update the PR accordingly. Thanks for the reference!

DanielNoord reviewed (commented):
`_is_singledispatchmethod_registration` is quite complex unfortunately 😅  But with some changes I think we can get this merged :)

Akshay9715 commented:
Thanks @DanielNoord for the helpful review!
I’ve implemented all the suggested changes — removed unnecessary `getattr`, replaced with proper `isinstance` checks, used `safe_infer`, and moved `_is_singledispatchmethod_registration` to the same module as discussed.
Please let me know if any further adjustments are needed 🙂

Pierre-Sassoulas commented on pylint/checkers/base/basic_error_checker.py:
```diff
@@ -95,6 +95,71 @@ def x(self, value): self._x = value
     return False
 
 
+def _extract_register_target(dec: nodes.NodeNG) -> nodes.NodeNG | None:
+    """
+    If decorator `dec` looks like `@func.register(...)` or `@func.register`,
+    return the `func` target node (Name or Attribute). Otherwise return None.
+    """
+    if isinstance(dec, nodes.Call):
+        func_part = dec.func
+        if isinstance(func_part, nodes.Attribute) and func_part.attrname == "register":
+            return func_part.expr
+        return None
+
+    if isinstance(dec, nodes.Attribute) and dec.attrname == "register":
+        return dec.expr
+
+    return None
```
I've been contaminated by Marc and I see them everywhere now, sorry.

```suggestion
def _extract_register_target(decorator: nodes.NodeNG) -> nodes.NodeNG | None:
    match decorator:
        case nodes.Call(func=nodes.Attribute(attrname="register", expr=expr)):
            # Looks like `@func.register(...)`
            return expr
        case nodes.Attribute(attrname="register", expr=expr):
            # Looks like `@func.register`
            return expr
        case _:
            return None
```

Akshay9715 replied:
Thanks! The `match` version looks elegant. I will use it in future refactors.

Pierre-Sassoulas reviewed (approved):
Great work @Akshay9715 , just a pet peeves of mine after reviewing 50 match PR by Marc Mueller :)

Akshay9715 commented:
@DanielNoord @Pierre-Sassoulas 
The PR is currently blocked by `codecov/patch` (97.05% of diff; target 100%). Would you like me to add tests to raise coverage for that, or will maintainers handle merging/coverage exceptions for this case?
